server/pulp/plugins/file/distributor.py

```diff
@@ -173,11 +174,27 @@ def publish_repo_fast_forward(self, repo, publish_conduit, config):
                             unit_checksum_old_set.add(checksum)
                             if checksum not in unit_checksum_set:
                                 unit_over_path_map[checksum] = fields[0]
+                _logger.debug("%d items were in MANIFEST %s, which exists? %s." % (
+                              len(unit_checksum_old_set), metadata_filename,
+                              os.path.exists(metadata_filename)))
 
                 # Copy incremental files into publishing directories
                 checksum_absent_set = unit_checksum_set - unit_checksum_old_set
+                _logger.debug("Increasing %d units" % len(checksum_absent_set))
+
+                # If added too many units, then publish repo with force_full
+                max_increase_units = min(50000, len(units) / len(hosting_locations))
+                if len(checksum_absent_set) > max_increase_units:
+                    config_tmp = copy.deepcopy(config)
+                    if not config_tmp.override_config:
+                        config_tmp.override_config = {}
+                    config_tmp.override_config['force_full'] = True
+                    self._rmtree_if_exists(build_dir)
+                    return self.publish_repo(repo, publish_conduit, config_tmp)
```

Comment:
I'd probably have gone with raising an exception instead, that avoids having to rewrite override_config and overall I think the code path would be less convoluted.

I mean for example, instead of having:

`publish_repo` => `publish_repo_fast_forward` => `publish_repo` (with different config)

You would instead have:

`publish_repo` => try `publish_repo_fast_forward`, if it fails, continue normally

Like this:

```
    if not config.get("force_full", False):
        try:
          return self.publish_repo_fast_forward(repo, publish_conduit, config)
        except FastForwardUnavailable:
          _logger.debug("Fast-forward publish bailed out, continuing normally")

    # (...rest of the publish_repo code here as normal)
```

Probably less confusing than having a method call an inner method which might call back to the outer method again with different arguments.

Reply:
Thanks for your suggestion, updated it.